lib/dal/src/attribute/value/dependent_value_graph.rs

```diff
@@ -401,23 +511,139 @@ impl DependentValueGraph {
         // Walking to the root for every value discovered
         // in the loop above ensures we construct a single
         // DVU connected graph for every tree of dependencies.
-        for id in self.inner.all_ids() {
+        let all_ids: Vec<_> = self.inner.all_ids().collect();
+        for id in all_ids {
             let mut cursor = id;
-            while let Some(parent_av_id) = AttributeValue::parent_id(ctx, cursor).await? {
-                self.value_depends_on(parent_av_id, cursor);
-                cursor = parent_av_id;
+            while let Some(parent_av_id) =
+                AttributeValue::parent_id(ctx, cursor.attribute_value_id()).await?
+            {
+                let parent_value = DependentValue::AttributeValue(parent_av_id);
+                self.value_depends_on(parent_value, cursor);
+                cursor = parent_value;
             }
         }
 
         Ok(())
     }
 
+    async fn leaf_overlays_for_component(
+        &mut self,
+        ctx: &DalContext,
+        component_id: ComponentId,
+    ) -> AttributeValueResult<Arc<Vec<ComponentLeaf>>> {
```

Comment:
Why an Arc? who else has ownership over this?

Reply:
I did the Arc for two reasons: (1) I didn't want to clone the whole vec to get it out of the map (maps return a ref, but that ref would break some borrow checker rules for the way we use it and (2) it has to be an Arc, not an Rc, because of async. Arc good! Fast! (on x86 systems the "Relaxed" ordering atomics are the same instructions as the "non" atomic instructions so an Rc and Arc are going to be very close to equivalent in machine code, although the compiler will do less reordering of instructions)